Dependency joberrors

kingsleyzissou opened the pull request:
This pull request includes:
Reporting that there has been an issue with a dependency job does not provide enough information about the dependency
failure. It makes more sense to bubble up the original error since this will tell us the information we need. It also has the added benefit of ensuring consistency of the status codes for metrics.

- [ ] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`

Information regarding our GitLab pipeline (Schutzbot):

CI will not be ran automatically if WIP label is applied or the PR is in DRAFT state, instead only
a link will be provided to the pipeline which can then be triggered manually if desired. To run the
CI automatically either switch the PR to ready or apply WIP+test label.

Outside contributors need manual approval from one of the osbuild-composer maintainers.

Schutzbot will only be triggered if all Tests jobs in GitHub workflow succeed.
-->

ondrejbudai commented:
I'm not sure if this approach is correct. Generally speaking, a job can have multiple dependencies and each of them can fail. With this approach, we would have to have an array of `JobError`s which sounds complicated. Also, we already have a job (`koji-finalize`) that does some stuff (and thus can fail) even though its dependency failed - so we have to communicate its status back somehow.

Since the api servers are the ones who are responsible for queueing jobs in the correct order, I think they should be ultimately responsible for inspecting the whole job chain and cherry-picking errors from it.

Gundersanne commented:
> I'm not sure if this approach is correct. Generally speaking, a job can have multiple dependencies and each of them can fail. With this approach, we would have to have an array of `JobError`s which sounds complicated. Also, we already have a job (`koji-finalize`) that does some stuff (and thus can fail) even though its dependency failed - so we have to communicate its status back somehow.
> 
> Since the api servers are the ones who are responsible for queueing jobs in the correct order, I think they should be ultimately responsible for inspecting the whole job chain and cherry-picking errors from it.

Didn't consider the koji case in this. In that case is it fine to consider all dependency failure errors as 4xx in metrics? That way it'll only the dependency shows up (in case of 5xxs).

kingsleyzissou commented:
Okay, will have to re-think this a bit then, I also hadn't considered the koji side of things properly.

> Didn't consider the koji case in this. In that case is it fine to consider all dependency failure errors as 4xx in metrics? That way it'll only the dependency shows up (in case of 5xxs).

I guess technically if a dependency failed, it's bad input for the next job (my two-cents)

kingsleyzissou commented:
@Gundersanne @ondrejbudai  I dropped the commit for propagating the dependency errors. I'll do that in a follow up.
For the meantime I changed the status codes for the metrics when we encounter a dependency error (assuming we're happy to go ahead with that change).
